sync2/multipeer/synclist.go

```diff
@@ -0,0 +1,60 @@
+package multipeer
+
+import (
+	"container/list"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+)
+
+// syncList keeps track of recent full syncs and reports whether the node is synced, that
+// is, the specified number of syncs has happened within the specified duration of time.
+type syncList struct {
+	mtx          sync.Mutex
+	clock        clockwork.Clock
+	minSyncCount int
+	duration     time.Duration
+	syncs        list.List
+}
+
+func newSyncList(clock clockwork.Clock, minSyncCount int, duration time.Duration) *syncList {
+	return &syncList{
+		clock:        clock,
+		minSyncCount: minSyncCount,
+		duration:     duration,
+	}
+}
+
+func (sl *syncList) prune(now time.Time) {
+	t := now.Add(-sl.duration)
+	for sl.syncs.Len() != 0 {
+		el := sl.syncs.Back()
+		if t.After(el.Value.(time.Time)) {
+			sl.syncs.Remove(el)
```

Comment:
can this not cause a memory leak? i.e. if a a series of double-linked list items get cut off from the rest, it means they might just continue living in memory because they reference each other (won't show up in the gc mark-and-sweep runs).

Reply:
Go's double-linked implementation unlinks the list element properly: https://github.com/golang/go/blob/go1.23.2/src/container/list/list.go#L108-L115